### PR TITLE
cdc:ignore timeout if snapshot is not complete

### PR DESCRIPTION
**Pull request by subodh1810**
Ref : https://airbytehq-team.slack.com/archives/C046N0TGRMX/p1667552437922099?thread_ts=1667481029.484859&cid=C046N0TGRMX

This change allows to ignore the `SUBSEQUENT_RECORD_WAIT_TIME` timeout if a snapshot is going on.

### CODE REVIEW

File: airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumRecordIterator.java [hunk 1]
```diff
@@ -80,9 +82,12 @@ protected ChangeEvent<String, String> computeNext() {
       // if within the timeout, the consumer could not get a record, it is time to tell the producer to
       // shutdown.
       if (next == null) {
-        LOGGER.info("Closing cause next is returned as null");
-        requestClose();
+        if ((!receivedFirstRecord || hasSnapshotFinished || maxInstanceOfNoRecordsFound >= 10) && !signalledClose) {
+          LOGGER.info("Closing cause next is returned as null");
```
Comment (akashkulk) [inline airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumRecordIterator.java:86]: Can you add a comment or add a better log message here to reflect the new logic?

**Review by akashkulk — APPROVED** [review]

File: airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumRecordIterator.java [hunk 2]
```diff
@@ -46,6 +46,7 @@ public class DebeziumRecordIterator extends AbstractIterator<ChangeEvent<String,
   private boolean receivedFirstRecord;
   private boolean hasSnapshotFinished;
   private boolean signalledClose;
+  private int maxInstanceOfNoRecordsFound;
```
Comment (rodireich) [inline airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumRecordIterator.java:49]: Adding heartbeats make it unnecessary but since only postgres has heartbeats for now I'll make sure it doesn't collide with heartbeats

File: airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCdcTargetPosition.java [hunk 3]
```diff
@@ -51,14 +51,20 @@ static PostgresCdcTargetPosition targetPosition(final JdbcDatabase database) {
 
   @Override
   public boolean reachedTargetPosition(final JsonNode valueAsJson) {
+    final SnapshotMetadata snapshotMetadata = SnapshotMetadata.valueOf(valueAsJson.get("source").get("snapshot").asText().toUpperCase());
     final PgLsn eventLsn = extractLsn(valueAsJson);
 
-    if (targetLsn.compareTo(eventLsn) > 0) {
+    if (SnapshotMetadata.TRUE == snapshotMetadata) {
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCdcTargetPosition.java:57]: Previously we initially looked at event `lsn` and if it reached target we returned `true` iso snapshotMetadata was `FALSE` or `LAST`.
Can you explain why this change?

Comment (subodh1810) [reply airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresCdcTargetPosition.java:57]: This change is mainly to denote that we dont trigger shutdown while snapshot is going on. there are 3 possible values for `snapshotMetadata` (`TRUE` , `FALSE` and `LAST`). TRUE means that snapshot is going on. LAST means that the record is the last value of snapshot and FALSE means that we are processing the transaction logs.

File: airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumRecordIterator.java [hunk 4]
```diff
@@ -80,9 +82,12 @@ protected ChangeEvent<String, String> computeNext() {
       // if within the timeout, the consumer could not get a record, it is time to tell the producer to
       // shutdown.
       if (next == null) {
-        LOGGER.info("Closing cause next is returned as null");
-        requestClose();
+        if ((!receivedFirstRecord || hasSnapshotFinished || maxInstanceOfNoRecordsFound >= 10) && !signalledClose) {
```
[the remark this answers is not in the document]
Comment (subodh1810) [reply airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumRecordIterator.java:85]: So the thing that happened with carcru's connection. Null was returned during snapshot cause Debezium took more time that 1 minute. We should have kept trying instead of shutting down

[the remark this answers is not in the document]
Comment (rodireich) [reply airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumRecordIterator.java:85]: This should hopefully be addressed by heartbeats

**Review by rodireich — APPROVED** [review]
